### ISSUE TITLE
On Windows: Kubernetes filter cannot find CA file

### ISSUE THREAD

**Issue by krancour** [fluent/fluent-bit#1838]
## Bug Report

On Windows only, the Kubernetes filter is failing to locate the CA file even though it exists at the specified path.

**To Reproduce**

Use @fujimotos [Windows branch](https://github.com/fluent/fluent-bit/tree/win32-next).

Configuration:

```
[FILTER]
        Name                kubernetes
        Match               kube.*
        Kube_URL            https://kubernetes.default.svc:443
        Kube_CA_File        /var/run/secrets/kubernetes.io/serviceaccount/ca.crt
        Kube_Token_File     /var/run/secrets/kubernetes.io/serviceaccount/token
        Kube_Tag_Prefix     kube.var.log.containers.
        Merge_Log           On
        Merge_Log_Key       log_processed
        K8S-Logging.Parser  On
        K8S-Logging.Exclude Off
```

Note this configuration works fine on the Linux nodes of my Kubernetes cluster.

Also note the following configuration with more Windows-looking paths yields the same error:

```
[FILTER]
        Name                kubernetes
        Match               kube.*
        Kube_URL            https://kubernetes.default.svc:443
        Kube_CA_File        C:\var\run\secrets\kubernetes.io\serviceaccount\ca.crt
        Kube_Token_File     C:\var\run\secrets\kubernetes.io\serviceaccount\token
        Kube_Tag_Prefix     kube.var.log.containers.
        Merge_Log           On
        Merge_Log_Key       log_processed
        K8S-Logging.Parser  On
        K8S-Logging.Exclude Off
```

Logs:

```
Fluent Bit v1.3.0
Copyright (C) Treasure Data

[2019/12/17 20:56:19] [ info] [storage] initializing...
[2019/12/17 20:56:19] [ info] [storage] in-memory
[2019/12/17 20:56:19] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2019/12/17 20:56:19] [ info] [engine] started (pid=4812)
[2019/12/17 20:56:19] [ info] [filter_kube] https=1 host=kubernetes.default.svc port=443
[2019/12/17 20:56:19] [error] [io_tls] flb_io_tls.c:165 X509 - Read/write of file failed
[2019/12/17 20:56:19] [error] [TLS] error reading certificates from /var/run/secrets/kubernetes.io/serviceaccount/ca.crt
[2019/12/17 20:56:19] [ info] [filter_kube] local POD info OK
[2019/12/17 20:56:19] [ info] [filter_kube] testing connectivity with API server...
[2019/12/17 20:56:19] [ warn] [filter_kube] could not get meta for POD fluent-bit-windows-2t9kv
[2019/12/17 20:56:19] [ info] [sp] stream processor started
```

**Expected behavior**

The filter should be able to locate the CA file.

**Your Environment**

Mixed node (Linux and Windows) AKS cluster.

**Additional context**

Just trying to successfully export logs from Windows nodes in my Kubernetes cluster.

cc @fujimotos

**Comment by krancour** [comment]
Note, the specified path is a symlink-- which I suspect is a factor.

**Comment by krancour** [comment]
@fujimotos, I pinned down the cause of this.

It seems your branch is pretty far behind master and contains a bug (a disagreement over number and order of params expected by the `flb_tls_context_new(...)` function) that was introduced by commit `dd405e008463f0696c831e2855d581d05fb64ba3`, but it does _not_ include the fix that was was introduced by `81ab2ac8a01bcfe09aa41d3bd8eded22b79c7482`.

Would a rebase to master, perhaps be in order here? Alternatively, it seems like there's a lot of changes in this branch and you might be setting yourself up for a difficult merge. Would it be worth starting to merge discrete changes that have been verified into master instead of continuing to diverge further from master?

**Comment by krancour** [comment]
After this is resolved, two other issues, pop up immediately. I'll open new issues for them, but they don't manifest until this is cleared, so I'm holding off for now and leaving a comment here so I don't lose track of it...

As @titilambert has noted in #960 there's an issue of `kubernetes.default.svc` not resolving. `kubernetes.default.svc.cluster.local` can resolve, but only if you sleep for a while first. This seems to be a timing issue related to DNS or other network infra.

```
[2020/01/04 03:08:14] [ info] [filter_kube] testing connectivity with API server...
[2020/01/04 03:08:15] [ warn] net_tcp_fd_connect: getaddrinfo(host='kubernetes.default.svc.cluster.local'): No such host is known. 
[2020/01/04 03:08:15] [error] [filter_kube] upstream connection error
```

If you clear _that_ issue, there's an issue (again, only on Windows) of getting a 400 (bad request) while testing connectivity with the server. The reason is that the HTTP `Host` header is not set:

```
[2020/01/03 23:31:54] [ info] [filter_kube] testing connectivity with API server...
[2020/01/03 23:31:54] [debug] [filter_kube] API Server (ns=brigade, pod=brigade-fluent-fluent-bit-windows-7gnbw) http_do=0, HTTP Status: 400
[2020/01/03 23:31:54] [debug] [filter_kube] API Server response
400 Bad Request: missing required Host header
[2020/01/03 23:31:54] [ warn] [filter_kube] could not get meta for POD brigade-fluent-fluent-bit-windows-7gnbw
```

Digging into the code, the `Host` header is conditionally set and that condition is never met, even on Linux. This doesn't seem to be a problem in Linux, however, and only on Windows. I don't know the kernel well enough to say this with certainty, but it may be possible this issue doesn't manifest on Linux because the kernel sets the `Host` header itself if not otherwise specified. (Maybe?) At any rate, it seems we should probably set this correctly regardless of OS.

**Comment by krancour** [comment]
> I don't know the kernel well enough to say this with certainty, but it may be possible this issue doesn't manifest on Linux because the kernel sets the Host header itself if not otherwise specified. (Maybe?)

Another possibility is that the 400 isn't coming from the API, but from the kube-proxy or some other bit of k8s networking that behaves differently on Windows nodes than on Linux nodes, i.e. requiring that header in the case of Windows but not otherwise. I'll try to see what I can find i applicable logs.

But I still think we should always set the Host header regardless.

**Comment by fujimotos** [comment]
@krancour Thanks for digging that up. 

> It seems your branch is pretty far behind master and contains a bug (a disagreement over number and order of params expected by the flb_tls_context_new(...) function) that was introduced by commit `dd405e008463f0696c831e2855d581d05fb64ba3`, but it does not include the fix that was was introduced by `81ab2ac8a01bcfe09aa41d3bd8eded22b79c7482`.

Yes, I can confirm this is the case.

To address the issue, I have just pulled the stable fixes into `win32-next`,
so it should be resolved on the latest tip of the branch.

> But I still think we should always set the Host header regardless.

I agree that we should set `Host` header here (namely, at line 223 of
kube_meta.c). Since HTTP/1.1 explicitly requires a `Host` header per spec,
I think we need to set it on all occasions on any platforms.

> Another possibility is that the 400 isn't coming from the API, but from the kube-proxy or some other bit of k8s networking that behaves differently on Windows nodes than on Linux nodes, i.e. requiring that header in the case of Windows but not otherwise. I'll try to see what I can find i applicable logs.

I suspect that the error occurs depending on golang's version used, rather
than the platform where k8s is running.

 * The error message is coming from golang's HTTP server implementation.
   In particular, `server.go` checks whether a request contains any Host
   header, and it aborts immediately if none found.

   https://go.googlesource.com/go/+/refs/heads/master/src/net/http/server.go#984

 * The particular failure path was added in 6e11f45ebd, which was released
   in golang v1.6 (on Feb 2016).

   https://go-review.googlesource.com/c/go/+/17892

I see several users reporting a similar issue (#1836 and #1850). Probably we
should set the header by default in flb_http_client.c.

**Comment by krancour** [comment]
@fujimotos thank you... I see you merged `v1.3.5` into the `win32-next` branch and when I build from that branch, everything works (assuming I use a startup script that sleeps for a bit while network issues work themselves out).

i.e. The issue involving the `Host` header is _not_ present in the `win32-next` branch at the moment, _however_ if I merge in `master` (or rebase to `master`-- whichever) the issue reappears.

I've tracked it down the `3fce26dce40bed4a9031ab48a7560e9708b3ec15`, which is a commit that is in `master` but _not_ in `v1.3.5`.

So... I think I should reiterate the perils I perceive in having this long-running feature branch that has diverged significantly from master. Not only are we missing bug fixes (as I was previously complaining) we are _also_ missing _new bugs_-- thereby giving us a false sense that what's in this branch is correct.

Can I again suggest that it might be time to start merging discrete PRs into `master` instead of continuing to work from the `win32-next` branch?

**Comment by fujimotos** [comment]
@krancour 

> Can I again suggest that it might be time to start merging discrete PRs
> into master instead of continuing to work from the win32-next branch?

Hmm. It seems that I need to rethink how to manage win32-next branch.
There are two different models on how the branch should work:

1. It should hold Windows patches on top of the stable release.
2. It should hold Windows patches on top of the unstable HEAD (master)

So far, I was managing win32-next as (1). This is because I wanted to
test *only* our Windows patches. Namely, I feared that it would muddle
our testing process if I incorporating unrelated changes.

("Is it broken due to the Windows patches, or other unrelated commits?")

... But I might be wrong. As you pointed out, for an integration testing,
it makes more sense if it contains all the changes for the next release,
since, by doing so, we are able to see what the Windows build _really_
looks like in the next stable release.

So in conclusion, I changed my mind and started to merged master into win32-next.

* https://github.com/fluent/fluent-bit/commits/win32-next
* https://ci.appveyor.com/project/fluent/fluent-bit-2e87g/builds/29940060

This is the latest build that contains Windows patches on top of master.
I hope it helps you.

**Comment by fujimotos** [comment]
Some addendum:

> i.e. The issue involving the Host header is not present in the win32-next branch at the moment, however if I merge in master (or rebase to master-- whichever) the issue reappears.
> I've tracked it down to 3fce26dce40bed4a9031ab48a7560e9708b3ec15, which is a commit that is in master but not in v1.3.5.

Yeah, my initial guess was wrong. The root cause was that k8s plugin
started to send HTTP requests without a Host header since 3fce26dce.

I posted a fix for the issue to #1858. I appreciate if you can confirm that
it actually fixes the issue...

> kubernetes.default.svc.cluster.local can resolve, but only if you sleep for a while first. This seems to be a timing issue related to DNS or other network infra.

Hmm. This seems to be very annoying issue. Do you have any idea how
we should fix this?

I'm guessing that this is related to kubernetes/kubernetes/issues/78479.

**Comment by krancour** [comment]
> So in conclusion, I changed my mind and started to merged master into win32-next.

This feels like a step in the right direction @fujimotos. **Thank you so much.** I'm still not entirely clear on the necessity of the `win32-next` branch in the first place, though. It seems the prevailing wisdom in the open source community is that smaller, focused, incremental changes are more manageable. i.e. If you're adequately convinced that certain individual changes in this branch are correct, shouldn't they be making their way back into master one by one?

> Yeah, my initial guess was wrong.

So were my first two guesses. 🤷‍♂ Just glad we found it.

> I posted a fix for the issue to #1858. I appreciate if you can confirm that
it actually fixes the issue...

Happy to help!



>> kubernetes.default.svc.cluster.local can resolve, but only if you sleep for a while first. This seems to be a timing issue related to DNS or other network infra.
>
> Hmm. This seems to be very annoying issue. Do you have any idea how
we should fix this?

Right now I use a startup script that sleeps for 60 seconds. That's ok for my particular use case, but I would never propose that as a universally acceptable solution.

I can ask around internally about any progress on kubernetes/kubernetes/issues/78479.

**Comment by krancour** [comment]
I commented in #1858, but I tested and it lgtm.

**Comment by krancour** [comment]
I'm also going to close this issue, because I think we're past anything that was discussed herein.
